### PR TITLE
Runtime.serial_bytes_available: return count instead of bool

### LINKED ISSUE #9031: runtime.serial_bytes_available returns boolean documentation indicates int
### CircuitPython version

```python
Adafruit CircuitPython 9.0.0-alpha.6-1-gfcfe2018bd on 2023-12-14; LILYGO T-DECK with ESP32S3
Adafruit CircuitPython 9.0.0-rc.0-dirty on 2024-03-08; M5 Stack Cardputer with ESP32S3
Adafruit CircuitPython 8.1.0-beta.2 on 2023-04-26; VCC-GND YD-ESP32-S3 (N16R8) with ESP32S3
Adafruit CircuitPython 8.2.9-9-g325ae7c6fd on 2024-02-13; Pimoroni Pico dv Base with rp2040
```

### Code/REPL

```python
print(type(runtime.serial_bytes_available))
```

### Behavior

```<class 'bool'>```

### Description

I thought this function returned the integer number of bytes available at one point, however all the boards/versions I've checked seem to return a simple boolean.

I'm not sure it makes sense to change the functionality as it's possible it could be a breaking change and the current behavior has apparently been in place since 8.x but then the documentation should probably be updated.

```
serial_bytes_available: int
Returns the whether any bytes are available to read on the USB serial input. Allows for polling to see whether to call the built-in input() or wait. (read-only)
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by dhalbert**
- Fixes #9031.

Change `supervisor.Runtime.serial_bytes_available` to return the number of available bytes instead of just a True or False. Since `0` is `False`-y, this works in the common case of `if supervisor.runtime`.serial_bytes_available`:`.

I looked in the Learn Guides and I only found one use of `.serial_bytes_available`, and it's exactly as above.

This was a much larger change than one might think because I did a refactor to reduce a bunch of identical `common-hal/supervisor` files to single files in `supervisor/shared`. So there is, net, a lot less code now.

### CODE REVIEW

File: ports/espressif/supervisor/usb_serial_jtag.c [hunk 1]
```diff
@@ -121,8 +121,8 @@ char usb_serial_jtag_read_char(void) {
     return c;
 }
 
-bool usb_serial_jtag_bytes_available(void) {
-    return ringbuf_num_filled(&ringbuf) > 0 || usb_serial_jtag_ll_rxfifo_data_available();
+uint32_t usb_serial_jtag_bytes_available(void) {
+    return ringbuf_num_filled(&ringbuf) || usb_serial_jtag_ll_rxfifo_data_available();
```
Comment (tannewt) [inline ports/espressif/supervisor/usb_serial_jtag.c:125]: This doesn't make sense to me. Shouldn't it be `+`?

File: supervisor/shared/serial.c [hunk 2]
```diff
@@ -260,52 +260,71 @@ char serial_read(void) {
...
-bool serial_bytes_available(void) {
+uint32_t serial_bytes_available(void) {
     #if CIRCUITPY_USB_VENDOR
-    if (tud_vendor_connected() && tud_vendor_available() > 0) {
-        return true;
+    if (tud_vendor_connected()) {
+        uint32_t count = tud_vendor_available();
+        if (count > 0) {
+            return count;
+        }
```
Comment (tannewt) [inline supervisor/shared/serial.c:269]: I think you probably want to add these all together too because you can read them all.

**Review by tannewt — CHANGES_REQUESTED** [review]
A couple functional questions. Thanks for cleaning this up!

File: ports/espressif/supervisor/usb_serial_jtag.c (hunk 1 above)
Comment (dhalbert) [reply ports/espressif/supervisor/usb_serial_jtag.c:125]: Yes, I thought the second thing was a boolean, but it is not.

File: supervisor/shared/serial.c (hunk 2 above)
Comment (dhalbert) [reply supervisor/shared/serial.c:269]: Aha, that makes sense -- I was not thinking of multiple available inputs.

**Review by tannewt — APPROVED** [review]
Thank you!

**Pull request merged**